### PR TITLE
Merkle Patricia Tree partial refactor

### PR DESCRIPTION
**Pull request by kumavis**

### CODE REVIEW

**Comment by kumavis** [comment]
so fresh and so clean

**Comment by kumavis** [comment]
all tests pass :smirk_cat: 

due to time constraints, did not refactor `updateNode` or `deleteNode` to use new `walkTrie` method. :crying_cat_face: 

committing a final checkpoint minimally dumps the scratch to the db :smile_cat: 

breaking changes: :scream_cat: 
`trie.db` was renamed `trie._db`
`trie.copy()` produces a trie that is independant from the original (checkpoint/commit/revert is not connected), but shares the db + scratch
I think thats all...

File: viz.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+      })
+    }, function(err, children){
+      nodeData.children = children
+      data[hash] = nodeData
+      walkController.next()
+    })
+  }, function () {
+    cb(null, data)
+  })
+}
```
Comment (wanderer) [inline viz.js]: So i love the Viz, but I would rather not have it part of this module directly. Maybe standlone module?

File: trieNode.js [hunk 2]
```diff
@@ -204,18 +240,22 @@ internals.nibblesToBuffer = internals.TrieNode.nibblesToBuffer = function (arr)
...
  * - branch - if the node is a branch
  * - extention - if the node is an extention
- * - unkown - if somehting fucked up
+ * - unknown - if somehting fucked up
  */
-internals.getNodeType = internals.TrieNode.getNodeType = function (node) {
+function getNodeType(node) {
   if (Buffer.isBuffer(node) || typeof node === 'string' || node instanceof String) {
-    return 'unkown';
+    return 'unknown';
```
Comment (wanderer) [inline trieNode.js]: can this ever be a case?

**Comment by wanderer** [comment]
cool looking good! I have two things to do on ethereumjs-lib that are blocking ATM
1) change copy in VM
2) check StateRoot hash in tests
After That I can test and merge

File: trieNode.js (hunk 2 above)
Comment (kumavis) [inline trieNode.js]: i dont think so

File: viz.js (hunk 1 above)
Comment (kumavis) [inline viz.js]: yep makes sense to me
i can delete it

File: baseTrie.js [hunk 3]
```diff
@@ -0,0 +1,662 @@
...
+
+
+function Trie(db, root) {
+
+  this.EMPTY_TRIE_ROOT = EMPTY_RLP_HASH;
+
+  this.sem = semaphore(1);
+
+  // setup dbs
+  this._db = db || levelup('', { db: memdown });
```
Comment (wanderer) [inline baseTrie.js]: what do you think about putting the db on `this.db` ?

**Comment by wanderer** [comment]
This example fails https://gist.github.com/wanderer/c4040f3a7df0b9298539 But doesn't fail in the old

**Comment by wanderer** [comment]
ok I made that example into a test on the master branch. https://github.com/wanderer/merkle-patricia-tree/blob/master/test/failingRefactorTests.js Ill another take a stab at it tomorrow

**Comment by wanderer** [comment]
so `checkpoint()` is failing somewhere

**Comment by wanderer** [comment]
ok so that is my fault, I changed copies behavior which is causing createScratchStream to fail

**Comment by wanderer** [comment]
:+1

**Pull request merged**